NGS2_WITNESS_Cycle1_experiment1_prep.R

```diff
@@ -0,0 +1,422 @@
+    }
+    return(tmp)
+}
+
+read_input_files <- function(dir) {
+    # identifies all csv files in a directory, reads then into r, and appends them
+    res <- list()
+    files <- list.files(dir, pattern = '.csv$')
+    sources <- strsplit(files, split = '_')
+    for(i in 1:length(files)) {
+        tmp <- read.csv(paste(dir, files[i], sep = '/'), header = TRUE,
+                        sep = ',', stringsAsFactors = FALSE)
+        tmp$source <- paste(sources[[i]][1:2], collapse = '_')
+        res[[i]] <- tmp
+        names(res)[i] <- paste(sources[[i]][1:2], collapse = '_')
+    }
+    return(res)
+}
+
+# Load raw breadboard data
+exp1 <- read_input_files(paste0('NGS2-Cycle', CYCLE, '-Experiment1/data'))
+
+# gather empanelment/breadboard crosswalk information
+empanel_bb_xwalk <- build_empanelment_bb_xwalk('oms_url_upload_20170821_1145.txt')
+
+# EXPERIMENT 1
+# cooperation dataset
+# create subsets for manipulation
+conn <- create_subsets(exp1, 'Connected', 'player', strip_practice = FALSE)
+act <- create_subsets(exp1, 'cooperationEvent', 'pid|round',
+                      strip_practice = TRUE)
+chng <- create_subsets(exp1, 'rewiringEvent', 'pid|nid|round',
+                       strip_practice = TRUE)
+chng <- lapply(chng, function(x) {
+    if(nrow(x) > 0) {
+        x$action_num <- ifelse(x$action == 'maintainConnection', 0,
+                               ifelse(x$action == 'breakConnection', -1, 1))
+    }
+    return(x)
+})
+
+# start building the `cooperation` output file
+coop <- build_cooperation(act)
+
+# calculate actor moves by round
+round_changes <- changes_by_round(chng, conn)
+nbr_connections_round <- calculate_connections(round_changes)
+
+# finish off the `cooperation` dataset
+coop <- mapply(function(x, y, exp) {
+    if(nrow(y) > 0) {
+        return(merge(x, y, by = c('round', 'pid')))
+    } else {
+        conns <- dcast(subset(exp, event == 'Connected'), id ~ data.name,
+                       value.var = 'data.value')
+        x$num_neighbors <- NA
+        for(i in 1:max(x$round)) {
+            x <- get_neighbor_number(x, conns, i)
+        }
+        return(x[, c('round', 'pid', 'action', 'group_size',
+                     'previous_decision', 'num_neighbors')])
+    }
+}, coop, nbr_connections_round, exp1, SIMPLIFY = FALSE)
+
+# add in parameters from dataset for completeness
+coop <- do.call(rbind, mapply(function(x, name, exp) {
+    exp_cond <- as.numeric(exp$data.value[exp$event == 'initParameters' &
+                                          exp$data.name == 'k'])
+    x$session <- name
+    x$condition <- ifelse(exp_cond == 0, 'Static',
+                          ifelse(exp_cond == .1, 'Viscous',
+                          ifelse(exp_cond == .3, 'Fluid',
+                          ifelse(exp_cond == 1, 'Random', 'Other'))))
+    return(x)
+}, coop, as.list(names(coop)), exp1, SIMPLIFY = FALSE))
+
+# add in empanelment id for non-bots
+coop <- merge_with_empanelment(coop, empanel_bb_xwalk, 'pid')
+
```

Comment:
Can we drop useless sessions here? I suggest the following code. 

# Drop failed "e1.1_2017-08-21" session, including "other" conditions
exp1_cooperation <- subset (exp1_cooperation, condition!="Other")
exp1_rewire <- subset (exp1_rewire, session!="e1.1_2017-08-21")

# Drop tiny "e1.4_2017-08-31" session, with only three players
exp1_cooperation <- subset (exp1_cooperation, session!="e1.4_2017-08-31")
exp1_rewire <- subset (exp1_rewire, session!="e1.4_2017-08-31")

Reply:
Happy to drop bad sessions, but would like to make it condition-based. The `condition == 'Other'` is fine, but for the other, it would be good to have some cutoffs... otherwise, we'll have to go through each experiment to determine if it should be dropped. That is, what's the minimum number of rounds and the minimum number of players you'll take as floors? From there, I'll work in the code to adjust based on those values.

Reply:
Minimum number of rounds = 2
Minimum number of players = 10

Reply:
This has been addressed with commit 8e90ef.